### ISSUE TITLE
equivalent of torch.load from io stream to gpu

### ISSUE THREAD

**Issue by vladmandic** [huggingface/safetensors#200]
sorry if i'm missing something, but i can't get things to work with `safetensors`  

goal is simple - provide loading progress bar while large model is loading and still load it to designated device

with `torch`, i can do:

```
with rich.progress.open(checkpoint_file, 'rb') as f:
  buffer = io.BytesIO(f.read())
  m = torch.load(buffer, map_location=device)
```

and with `safetensors`, i have a choice of:

```
m = safetensors.torch.load_file(checkpoint_file, device=device)
```
but this only works with filename, it doesn't take buffer as argument

or i can do 
```
with rich.progress.open(checkpoint_file, 'rb') as f:
  buffer = f.read()
  m = safetensors.torch.load(buffer)
```
but then i cannot specify `device` so model attempts to load on cpu and fails with oom
also, input is not `bytes` like it is with `torch`, but thats minor

**Comment by Narsil** [comment]
Hi.

Currently not we cannot accept file-like inputs, and the reason is simple, it's using memory mapping (or torch.Storage) under the hood, which cannot work with file-like objects.

Have you checked how long it takes to do 
```
weights = safetensors.torch.load_file(checkpoint_file, device="cpu")
```
It should be almost instant regardless of the size of the file (this it's not allocating anything.

You can then add some progressbar when sending to GPU for instance

```
weights = {k: v.to('cuda:0') for k, v in progressbar(weights.items())}
```

Would that work ?

**Comment by vladmandic** [comment]
a lot more testing - seems to me like `safetensors.torch.load_file` is just bad while `safetensors.torch.load` is muuuuch better and far more consistent

loading same ~4gb diffusion model from `ckpt` and `safetensors`:

|op|ckpt load to cpu|ckpt load to cuda|safetensors load_file to cpu|safetensors load_file to cuda|safetensors load|code|
|---|---|---|---|---|---|----|
|load weights|6.4s|6.0s|0.9s|66.2s|5.5s| (see below)
|create model|0.3s|0.3s|0.3s|0.3s|0.3s| `model = instantiate_from_config(config)`
|apply weights|0.4s|1.0s|18.7s|1.0s|0.6s| `model.load_state_dict(weights)`
|device move|0.7s|0.5s|0.5s|0.5s|0.8s| `model.to('cuda')`

loading code:
```python
# ckpt load
with open(checkpoint_file, 'rb') as f:
  buffer = f.read()
  weights = safetensors.torch.load(buffer)

# safetensors load_file
weights = safetensors.torch.load_file(weights_file, device=xxx)

# safetensors load
with open(checkpoint_file, 'rb') as f:
  buffer = io.BytesIO(f.read())
  pl_sd = torch.load(buffer)
```

overall, `safetensors.torch.load_file` is **2x** (best) to **10x** (worst) slower than `ckpt` for loading a model  
and it spends time in a totally different part of the code:
- `safetensors` load to cpu is majority during `load_state_dict` which includes moving selected weights to gpu  
  actual `load_file` is negligible as its only to initiate a mmap for further use  
- `safetensors` load to gpu is majority during `load_file` and its incredibly slow  

while `safetensors.torch.load` behaves very consistently and with very good performance  

btw, using suggested loop is incredibly slow - nearly **100sec** alone:
> weights = {k: v.to('cuda:0') for k, v in weights.items()}

**Comment by Narsil** [comment]
Can you share both files ? Including the entire test script ?
I would love to help figure this out but currently I am still unable to reproduce.

Additionally could you share:
- torch version
- OS version (Windows correct ?)
- safetensors version
- Disk type (HDD correct) ?

`0.9s` seems extremely slow for loading on CPU. It shouldn't be above 1ms ever (I need to recheck on Windows but still).

Given what you are sharing, I think most of the slowdown seems to be from sequential reads VS non sequential reads. With `safetensors.torch.load` you started by reading the entire file in one sequential READ, and then did whatever you wanted to do with it.
Please note, that while on HDD 1 sequential read will most likely be better, this is not the case for SSDs where sequential reads don't have the same performance profiles, and reading tensor per tensor does allow for lesser overall RAM usage (which could matter)

With `safetensors.torch.load_file` since you're memory mapping, I am guessing it's issuing read calls for each individual tensors (I have no clue why Windows might do that instead of just reading the entire file in memory).
It could be either the Windows implementation of `torch.Storage` or something to do with how Windows actually handles memory maps.

The performance profile you are sharing a super valuable and interesting, but currently I don't have a good clue as to why loading directly on GPU can be 3x slower than loading on CPU, and then on GPU.
Have you made sure in all of your profiling, to load the files at LEAST twice. That is to ensure, the actual disk IO is **not** taken into account ? Disks +OS go to extreme lengths to avoid reading, and will cache as much as they can, which can totally mess up benchmarks if you're benchmarking a cached file VS an actual disk read file.

If you really want to measure DISK IO, the only good reproduceable road I have found is to *reboot* between each benchmark. (It's excruciatingly painful to do, but I didn't find any other good solutions, since now lots of disks also have their own caching system which seems impossible to flush).

Another one is cuda initialization. `torch.zeros((2, 2)).cuda()` takes a really long time the first time it's called, since a lot of initialization has to be done there. Depending on the benchmark script that can screw a lot the measurements.


Another potential culprit could have to do with pointer alignment but this I cannot check without being able to reproduce.


Note: I took a look at your PR, you cannot remove `map_location` for `torch.load`. Torch pickled files contain the actual location of a tensor (so cuda:0 or cuda:1) and will attempt to load those on GPU even if user doesn't have those devices. In order to load appropriately on the device of the user, you need to specify `map_location` to avoid the files not being loadable on some systems.

**Comment by vladmandic** [comment]
Environment:
- Ubuntu 22.04 running in WSL2 inside Windows 11  
  storage hosting model files is NTFS formatted SSD  
  so going from Linux -> WSL -> Windows -> NTFS likely does cause some `mmap` overhead, but it does work  
  (in wsl2 windows exposes ntfs drives as drvfs loopback mounts to linux)  
  SSD itself is Samsung 980Pro, pretty much as fast as home drives get  
- Torch 2.0.0+cu118 CUDA 11.8 cuDNN 8700
- Safetensors 0.3.0

> Have you made sure in all of your profiling, to load the files at LEAST twice.

I've repeated all tests at least 3 times. Results do vary, but not by much, what I've posted is representative.

Anyhow, project itself is too massive to trace down all variables, I'll create a separate small one for testing/benchmarking purposes and will update here.

> Note: I took a look at your PR

PR was premature, I removed it, just ignore that part.

**Comment by Narsil** [comment]
Thanks, if you had a link to the files, that would be nice for reproduction too.

- Pointer alignment is a thing
- Tensor order is a thing (differences in disk read order maybe).

Both depend on the different files.
Also I know operation order in is really important with CUDA and the proposed change (load everything on CPU and then move to GPU) is actually something that I've noticed to be faster that loading on GPU all the tensors.
This on bare Linux.

Basically it does:
- Lazy load tensor x on CPU, Lazy load tensor y, ......, move x to cuda, move y to cuda
Instead of
- Lazy load tensor x, move x to cuda, lazy load tensor y, move y to cuda, ....

Somehow the first option is indeed faster than the second one

> so going from Linux -> WSL -> Windows -> NTFS likely does cause some mmap overhead, but it does work
> (in wsl2 windows exposes ntfs drives as drvfs loopback mounts to linux)

Good information to try and reproduce. Could definitely be something here.

**Comment by vladmandic** [comment]
> Thanks, if you had a link to the files, that would be nice for reproduction too

<https://huggingface.co/stabilityai/stable-diffusion-2-1/tree/main>  
contains both safetensors and ckpt version of the same model.

but...original `StableDiffusionPipeline` class is an opaque class that handles loads internally, so to reproduce model creation and moving weights, need to break it down and reimplement in a simple way to allow for manual loading experiments.

project i'm working on already does that, but it does sooo much more that it introduces too many variables to have valid reproducible benchmark.

**Comment by Narsil** [comment]
`diffusers`  or `stable-diffusion-webui` ? There are the 2 sets of weights here.

**Comment by vladmandic** [comment]
> `diffusers` or `stable-diffusion-webui` ? There are the 2 sets of weights here.

[v2-1_768-ema-pruned.ckpt](https://huggingface.co/stabilityai/stable-diffusion-2-1/blob/main/v2-1_768-ema-pruned.ckpt)
[v2-1_768-ema-pruned.safetensors](https://huggingface.co/stabilityai/stable-diffusion-2-1/blob/main/v2-1_768-ema-pruned.safetensors)

**Comment by Narsil** [comment]
@vladmandic 

First of all, thank you for the reproducible benchmark ! Very worthwile.

Tl;dr

- I am tempted to file a mmap bug upstream (need to confirm torch uses mmap with Storage on Windows) directly for Windows. It could be because NTFS. I also would like to check about fragmentation of those files to discard that possibility.
- I don't think we should make any changes in this library, since isolated loads seem to work really well, problem occurs only when reading disk poorly (which is extremely slow on torch too) and when torch has been fragmented already which I don't think this library is responsible for (we're declaring a contiguous Storage, so there's no better way. The general case still seems to be hot file (people tend to swap models regularly so most of the time the file should be hot), and the torch fragmentation affects both ckpt and safetensors, so it's not really linked to either files.



I have seen the humongous slowdowns you have seen, hinting that it's not WSL being the issue (which is good somehow).
However, problems seem to have vanished after a few runs. I'm still quite puzzled by it.
I think it could be linked to Windows + NTFS read ordering. (For that slurping the entire file is better).
But I was never was able to reproduce it after the first initial attempt. !!!????

Then when the issue was gone, I was consistently seeing safetensors being slower than torch, UNLESS I reversed the loading order, in which case the opposite was true.


This leads me to think there's some kind of **fragmentation** within torch memory arena. 
This is a behavior also seen on Linux 
Despite you unloading the model, torch decides to do whatever it wants with its memory, so some stuff can be in there, forcing subsequent tensors to cause reallocation in instead of fitting in memory, and hence slower allocation.
Usually `torch.cuda.empty_cache()` can help alleviate this issue, however in this case it seemed to send the model creation into impossibly long start. (I have no idea why it would cause this, I have never seen this happen on linux).


Then from a **cold start** (so file not in disk memory or OS cache):
- safetensors (the first one in this case) loads in ~100s (47s seconds for model creation and ~50s for moving the weights)
- ckpt in ~58s (18s for model creation which is more in line with subsequent runs, and ~38s for reading the file).
(So I would say we should only compare 50 to 38 to make it fair since both are cold reads, cold cuda also seems to be a thing)

My conclusion from this is that indeed the memory mapping reads in unordered, and seems to be slower than reading in ordered.

**However**, on subsequent runs I get faster loads on safetensors consistently:

SF:
- ~300ms to load the weights
- ~1,243s to move the weights

torch:
- ~4.5 to load the weights

See below for a more complete picture. For both ckpt and sf, the `file` loads faster than `stream` (by a small margin).

Also you can check the following code for a micro benchmark:

```python
from safetensors.torch import load_file, save_file
import os
import datetime
import torch

torch.zeros((2, 2)).cuda()


repeats = 1
checkpoint_folder = os.path.join("C:/Users", 'Administrator', 'Downloads') # update path with your location
checkpoint_files = [ # download files from (https://huggingface.co/stabilityai/stable-diffusion-2-1/tree/main)
    # 'v2-1_768-ema-pruned.ckpt',
    'v2-1_768-ema-pruned2.safetensors'
]


if __name__ == "__main__":
    checkpoint_file = os.path.join(checkpoint_folder, checkpoint_files[0])
    start = datetime.datetime.now()
    weights = load_file(checkpoint_file, device='cuda:0')
    print(f"Load on gPU {datetime.datetime.now() - start}")
    start = datetime.datetime.now()
    weights = load_file(checkpoint_file, device='cpu')
    print(f"Load on cPU {datetime.datetime.now() - start}")

    start = datetime.datetime.now()
    weights = load_file(checkpoint_file, device='cpu')
    weights = {k : v.to('cuda:0') for k , v in weights.items()}
    print(f"Load on cPU-> GPU {datetime.datetime.now() - start}")
```






(Hot run)
```
12:08:23-474843 torch: 1.13.1+cu117 cuda: 11.7 cudnn: 8500 gpu: Tesla T4 capability: (7, 5)                  bench.py:40
                start tests using file method                                                                bench.py:40
                load model v2-1_768-ema-pruned2.safetensors using file method pass 1 of 1 start              bench.py:40
12:08:23-647308 0.015 load config: v2-inference-768-v.yaml (cpu: 0.242 gpu: 0.099 gc: 0.154)                 bench.py:57
No module 'xformers'. Proceeding without it.
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
12:08:43-006666 19.28 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.233 gpu: 0.099 gc:     bench.py:57
                0.081)
12:08:43-366233 0.298 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned2.safetensors (cpu:  bench.py:57
                5.237 gpu: 0.099 gc: 0.063)
12:08:43-428807 0.0 state dict: 1242 (cpu: 5.237 gpu: 0.099 gc: 0.052)                                       bench.py:57
12:08:44-728099 1.243 apply weigths (cpu: 10.09 gpu: 0.099 gc: 0.065)                                        bench.py:57
12:08:45-502558 0.701 model eval (cpu: 7.663 gpu: 0.099 gc: 0.064)                                           bench.py:57
12:08:46-522329 0.95 model move (cpu: 5.23 gpu: 2.629 gc: 0.077)                                             bench.py:57
12:08:46-616546 0.003 model unload (cpu: 5.23 gpu: 0.099 gc: 0.091)                                          bench.py:57
12:08:47-281353 load model v2-1_768-ema-pruned2.safetensors using file method pass 1 of 1 in 23.803 seconds  bench.py:40
12:08:47-285511 start tests using stream method                                                              bench.py:40
12:08:47-288053 load model v2-1_768-ema-pruned2.safetensors using stream method pass 1 of 1 start            bench.py:40
12:08:47-360677 0.018 load config: v2-inference-768-v.yaml (cpu: 0.372 gpu: 0.099 gc: 0.064)                 bench.py:57
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
12:09:04-991196 17.492 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.234 gpu: 0.099 gc:    bench.py:57
                0.139)
12:09:09-616203 4.539 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned2.safetensors (cpu:  bench.py:57
                10.087 gpu: 0.099 gc: 0.081)
12:09:09-678674 0.0 state dict: 1242 (cpu: 10.087 gpu: 0.099 gc: 0.051)                                      bench.py:57
12:09:10-225540 0.484 apply weigths (cpu: 10.087 gpu: 0.099 gc: 0.065)                                       bench.py:57
12:09:11-069314 0.764 model eval (cpu: 7.664 gpu: 0.099 gc: 0.085)                                           bench.py:57
12:09:12-038044 0.889 model move (cpu: 5.232 gpu: 2.629 gc: 0.066)                                           bench.py:57
12:09:12-100546 0.003 model unload (cpu: 5.232 gpu: 0.099 gc: 0.065)                                         bench.py:57
12:09:12-319297 load model v2-1_768-ema-pruned2.safetensors using stream method pass 1 of 1 in 25.045        bench.py:40
                seconds
                start tests using file method                                                                bench.py:40
12:09:12-334921 load model v2-1_768-ema-pruned.ckpt using file method pass 1 of 1 start                      bench.py:40
12:09:12-397420 0.013 load config: v2-inference-768-v.yaml (cpu: 0.372 gpu: 0.099 gc: 0.061)                 bench.py:57
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
12:09:30-178901 17.701 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.234 gpu: 0.099 gc:    bench.py:57
                0.071)
12:09:32-647264 2.405 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned.ckpt (cpu: 10.087   bench.py:57
                gpu: 0.099 gc: 0.068)
12:09:32-710224 0.0 state dict: 1242 (cpu: 10.087 gpu: 0.099 gc: 0.058)                                      bench.py:57
12:09:33-311658 0.531 apply weigths (cpu: 10.087 gpu: 0.099 gc: 0.068)                                       bench.py:57
12:09:34-241392 0.817 model eval (cpu: 7.665 gpu: 0.099 gc: 0.111)                                           bench.py:57
12:09:35-194166 0.877 model move (cpu: 5.232 gpu: 2.629 gc: 0.075)                                           bench.py:57
12:09:35-287889 0.003 model unload (cpu: 5.232 gpu: 0.099 gc: 0.085)                                         bench.py:57
12:09:35-522346 load model v2-1_768-ema-pruned.ckpt using file method pass 1 of 1 in 23.197 seconds          bench.py:40
                start tests using stream method                                                              bench.py:40
                load model v2-1_768-ema-pruned.ckpt using stream method pass 1 of 1 start                    bench.py:40
12:09:35-616168 0.012 load config: v2-inference-768-v.yaml (cpu: 0.372 gpu: 0.099 gc: 0.066)                 bench.py:57
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
12:09:53-333501 17.651 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.235 gpu: 0.099 gc:    bench.py:57
                0.079)
12:09:58-179379 4.757 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned.ckpt (cpu: 10.088   bench.py:57
                gpu: 0.099 gc: 0.077)
12:09:58-227260 0.0 state dict: 1242 (cpu: 10.088 gpu: 0.099 gc: 0.049)                                      bench.py:57
12:09:58-830040 0.526 apply weigths (cpu: 10.088 gpu: 0.099 gc: 0.079)                                       bench.py:57
12:09:59-632039 0.727 model eval (cpu: 7.666 gpu: 0.099 gc: 0.069)                                           bench.py:57
12:10:00-622017 0.913 model move (cpu: 5.233 gpu: 2.629 gc: 0.068)                                           bench.py:57
12:10:00-694152 0.003 model unload (cpu: 5.233 gpu: 0.099 gc: 0.068)                                         bench.py:57
12:10:00-955889 load model v2-1_768-ema-pruned.ckpt using stream method pass 1 of 1 in 25.428 seconds
```

(Cold run)
```
2:01:37-975768 45.497 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.232 gpu: 0.099 gc:    bench.py:57
                0.135)
12:01:38-694403 0.635 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned2.safetensors (cpu:  bench.py:57
                5.236 gpu: 0.099 gc: 0.073)
12:01:38-756835 0.0 state dict: 1242 (cpu: 5.236 gpu: 0.099 gc: 0.066)                                       bench.py:57
12:02:30-033250 51.198 apply weigths (cpu: 10.089 gpu: 0.099 gc: 0.077)                                      bench.py:57
12:02:30-959694 0.843 model eval (cpu: 7.663 gpu: 0.099 gc: 0.077)                                           bench.py:57
12:02:31-904412 0.883 model move (cpu: 5.229 gpu: 2.629 gc: 0.072)                                           bench.py:57
12:02:31-983510 0.003 model unload (cpu: 5.228 gpu: 0.099 gc: 0.075)                                         bench.py:57
12:02:32-616182 load model v2-1_768-ema-pruned2.safetensors using file method pass 1 of 1 in 100.581 seconds bench.py:40
12:02:32-631812 start tests using stream method                                                              bench.py:40
                load model v2-1_768-ema-pruned2.safetensors using stream method pass 1 of 1 start            bench.py:40
12:02:32-709907 0.012 load config: v2-inference-768-v.yaml (cpu: 0.372 gpu: 0.099 gc: 0.068)                 bench.py:57
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
12:02:50-603859 17.825 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.235 gpu: 0.099 gc:    bench.py:57
                0.075)
12:02:55-424402 4.738 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned2.safetensors (cpu:  bench.py:57
                10.088 gpu: 0.099 gc: 0.076)
12:02:55-487906 0.0 state dict: 1242 (cpu: 10.088 gpu: 0.099 gc: 0.064)                                      bench.py:57
12:02:56-164621 0.598 apply weigths (cpu: 10.088 gpu: 0.099 gc: 0.07)                                        bench.py:57
12:02:56-954957 0.7 model eval (cpu: 7.665 gpu: 0.099 gc: 0.089)                                             bench.py:57
12:02:57-959698 0.936 model move (cpu: 5.233 gpu: 2.629 gc: 0.07)                                            bench.py:57
12:02:58-038252 0.003 model unload (cpu: 5.233 gpu: 0.099 gc: 0.074)                                         bench.py:57
12:02:58-256976 load model v2-1_768-ema-pruned2.safetensors using stream method pass 1 of 1 in 25.636        bench.py:40
                seconds
                start tests using file method                                                                bench.py:40
12:02:58-272685 load model v2-1_768-ema-pruned.ckpt using file method pass 1 of 1 start                      bench.py:40
12:02:58-335184 0.012 load config: v2-inference-768-v.yaml (cpu: 0.373 gpu: 0.099 gc: 0.062)                 bench.py:57
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
12:03:16-319553 17.894 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.235 gpu: 0.099 gc:    bench.py:57
                0.081)
12:03:54-312115 37.907 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned.ckpt (cpu: 10.088  bench.py:57
                gpu: 0.099 gc: 0.078)
12:03:54-375496 0.0 state dict: 1242 (cpu: 10.088 gpu: 0.099 gc: 0.063)                                      bench.py:57
12:03:54-851985 0.408 apply weigths (cpu: 10.088 gpu: 0.099 gc: 0.072)                                       bench.py:57
12:03:55-704567 0.785 model eval (cpu: 7.666 gpu: 0.099 gc: 0.065)                                           bench.py:57
12:03:56-662854 0.889 model move (cpu: 5.234 gpu: 2.629 gc: 0.071)                                           bench.py:57
12:03:56-740978 0.003 model unload (cpu: 5.234 gpu: 0.099 gc: 0.075)                                         bench.py:57
12:03:57-084998 load model v2-1_768-ema-pruned.ckpt using file method pass 1 of 1 in 58.824 seconds          bench.py:40
                start tests using stream method                                                              bench.py:40
12:03:57-100537 load model v2-1_768-ema-pruned.ckpt using stream method pass 1 of 1 start                    bench.py:40
12:03:57-178805 0.012 load config: v2-inference-768-v.yaml (cpu: 0.374 gpu: 0.099 gc: 0.07)                  bench.py:57
LatentDiffusion: Running in v-prediction mode
DiffusionWrapper has 865.91 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
12:04:14-928495 17.644 create model: ldm.models.diffusion.ddpm.LatentDiffusion (cpu: 5.235 gpu: 0.099 gc:    bench.py:57
                0.114)
12:04:19-459765 4.45 load weights: C:/Users\Administrator\Downloads\v2-1_768-ema-pruned.ckpt (cpu: 10.089    bench.py:57
                gpu: 0.099 gc: 0.078)
12:04:19-522264 0.0 state dict: 1242 (cpu: 10.089 gpu: 0.099 gc: 0.057)                                      bench.py:57
12:04:20-133472 0.534 apply weigths (cpu: 10.089 gpu: 0.099 gc: 0.074)                                       bench.py:57
12:04:20-955599 0.746 model eval (cpu: 7.667 gpu: 0.099 gc: 0.078)                                           bench.py:57
12:04:21-959797 0.915 model move (cpu: 5.234 gpu: 2.629 gc: 0.085)                                           bench.py:57
12:04:22-053891 0.003 model unload (cpu: 5.234 gpu: 0.099 gc: 0.081)                                         bench.py:57
12:04:22-272759 load model v2-1_768-ema-pruned.ckpt using stream method pass 1 of 1 in 25.176 seconds        bench.py:40
PS C:\Users\Administrator\src\sd-loader>
```

**Comment by vladmandic** [comment]
running in **wsl** has one advantage - i can restart a vm nearly instantly for repeatable results  
and i can test with different storage types (local or ntfs drive mounted via drvfs)  

so i've parametrized everything in my `bench.py` and re-did all the tests with vm restart between each test  
and results are mostly in-line with yours  

```log
./bench.py --help

options:
  --model MODEL             model name
  --method {file,stream}    load method
  --repeats REPEATS         number of repeats
  --device {cpu,cuda}       load initial target
  --config CONFIG           model config
  --dtype {fp32,fp16,bf16}  target dtype
```

and yes, by far the fastest load being `safetensors.load_file` with 1.5sec load+apply weights, but with a bit of a catch  

### tl;dr

- `safetensors.load_file` is *amazing* for local files when **mmap** is working and *horrible* for non-local files  
  it still tries to use **mmap** and its causing i/o storm with high cpu load, disk saturation and load read speads  
- `safetensors.load_file` to `cuda` is just broken when file is non-local  
- using `fp16` or `fp32` has massive impact on memory usage, but very little impact on load times  
- load to `cpu` and move to `cuda` is always faster than load to `cuda` (when its working at all)  

unrelated, but fyi, i've noticed that python variable scope has massive impact on torch gc  
e.g. using global variable will block torch gc even if variable is deleted

### using model stored on ntfs drive mounted via loopback drvfs

- `torch.load` using **file**:  
  - very consistent results  
    load **~10sec** regardless if loading to `cpu` or `cuda` or if using `fp16` or `fp32`  
    minimum effect of os buffering  
    disk reads 500mb+/sec with 25% saturation  
    cpu load: 10-20% of all-cores  
- `torch.load` using **stream**:  
  - load slightly slower at **~12sec**  
    some effect of os buffering as initial load is 50% longer at ~18sec  
  - system load same as if using file method  
- `safetensor.load_file` using **file**:  
  - load to `cuda` is non-viable, takes forever, aborted
  - load to `cpu` is 2sec **load** + 35sec **apply weigts**  
    this proves that **mmap** is actually working, but not efficient at all  
    disk reads ~150mb/sec with 85% saturation: there are so many io ops that reads are not even saturating the disk and cpu load: 50-60% of all-cores  
- `safetensor.load` using **stream**:  
  - load to `cuda` does not exist as device cannot be specified  
  - load to `cpu` is consistent and pretty much identical to `torch.load`  
  - system load is consistent and pretty much identical to `torch.load`  

### using model stored on local storage using ext4 fs

in all cases, disk reads hit peaks of 2gb+/sec with low saturation and low cpu load  

- `torch.load` using **file**:  
  - very consistent results at **~2.5sec** load  
    minimum effect of os buffering  
- `torch.load` using **stream**:  
  - load slightly slower at **~3sec**  
  - system load same as if using file method  
- `safetensor.load_file` using **file**:  
  - using `cpu` as target its incredibly fast, 0.2sec **load** + 1sec **apply weights**  
    disk reads hit peaks of 2gb+/sec  
  - using `cuda` as target results in slower ~4sec **load** + 0.5sec **apply weights**  
- `safetensor.load` using **stream**:  
  - slightly slower, but still very fast, 2.5sec **load**  
    some effect of os buffering as initial pass is 2x at ~5sec **load**  

### thoughts

wsl2 uses linux plan9 loopback virtual filesystem to mount ntfs drive which clearly has some major performance issues as even best case, disk reads are 1/3 of native disk reads (500-600mb/s vs 2+gb/s)  
but when used with **mmap** its horrible  

there are set of patches that may help, but i'm guessing it could be a while to see them in linux kernel and then in wsl2  
<https://lore.kernel.org/lkml/20221218232217.1713283-1-evanhensbergen@icloud.com/>

but...just because something is *compatible* with mmap doesn't mean it should be used  
mmap is always tricky and there should be some sanity checks when using it  

yes, the root issue here is "bad" filesystem, but i'm expecting same issue to happen anywhere where mmap is less than desirable - for example, when model file is stored on a network drive or in a cloud storage and mounted in a virtual filesystem - reality is that mmap causes far more i/o requests which on non-local filesystems have a major latency. so for anything except local storage, streaming mode is far more desirable  
this would possibly even apply to storage mounted using mediums using "slower" protocols, such as usb2 drives where protocol overhead on each i/o is not insignificant

**Comment by Narsil** [comment]
>  so for anything except local storage, streaming mode is far more desirable

Do you have any suggestions to make this work ? 
IMHO, a filename is supposed to be local storage, and local storage is 90% of the actual usage. (Both personal computers and cloud). If something else is posing as local storage, then there's little to be done about caveats.

It seems to me that in any case, the solution seems either by setting a flag within stable-diffusion-webui (stream or file mode).
Wdyt ?

**Comment by vladmandic** [comment]
> > so for anything except local storage, streaming mode is far more desirable
> 
> Do you have any suggestions to make this work ? IMHO, a filename is supposed to be local storage, and local storage is 90% of the actual usage. (Both personal computers and cloud). If something else is posing as local storage, then there's little to be done about caveats.
> 
> It seems to me that in any case, the solution seems either by setting a flag within stable-diffusion-webui (stream or file mode). Wdyt ?

Yes, that's exactly what I was thinking.
Alternatively, a bit more complicated to implement, but may scale better is to support multiple folders and have method set per-folder. E.g, I may want to have few models local and larger archive on non-local storage.

**Comment by brendanhoar** [comment]
Just another data point here, using both A1111 and ComfyUI (currently as SwarmUI backend), both with the safetensors library, on Windows 11...

Using Sysinternals Disk Monitor, I see that the data load pattern (on my windows system anyway) for reading data from a newly "seen" (that is, uncached) safetensors file is:

1. The default way of using A1111, which uses the **load_file** method, I see mostly 64KB or smaller physical IOs to the drive when accessing the file.
2. Enabling the A1111 option of  "_Disable memmapping for loading .safetensors files. (fixes very slow loading speed in some cases)_", which first reads the entire file into a bytearray and then passes it via the **load** method, I see 4MB physical IOs to the drive.
3. Similarly ComfyUI uses the default **load_file** method, which produces 64KB or smaller physical IOs to the drive.
4. If I patch** ComfyUI with the same change from the A1111 option (# 2 above), I see 4MB physical IOs to the drive.

The end result is that the load method produces on the order of 1600 OS-side physical IO requests with an SDXL model, whereas load_file produces on the order of ~100,000 physical IO requests with an SDXL model.

So, similar to the above conversations, load_file is excruciatingly slow.

Without further analysis nor understanding of the lower layer (here be dragons): 

I theorize what is happening is that the loading tensors onto the GPU from the source (whether it be a bytebuffer or a memory mapped file) results in a large number of (64KB?) DMA transactions, and when using the memory mapped file, that results in the same number of 64KB physical IO requests on my system (which also appear to be out of order?), leading to significant latency/overhead in the storage subsystem.  

Contemporary NVME PCIe SSDs can mask some of the problems with that approach, but the storage software layers may come into play as well: e.g. the drives I have safetensors files on are veracrypt partitions, and that may be exacerbating the issue (there's a known performance problem with certain types of access patterns with that software).

** path of utils.pya
`
    if ckpt.lower().endswith(".safetensors"):
        #-BGH 1. original comfy
        #-BGH sd = safetensors.torch.load_file(ckpt, device=device.type)
        #-BGH 2. A1111 alternate
        sd = safetensors.torch.load(open(ckpt, 'rb').read())
`

**Comment by vladmandic** [comment]
yup, that's the same underlying issue.
(and btw, "disable memmapping for loading .safetensors files" in a1111 is my code)

**Comment by brendanhoar** [comment]
> yup, that's the same underlying issue. (and btw, "disable memmapping for loading .safetensors files" in a1111 is my code)

And we all thank you for it. :)

**Comment by richardm1** [comment]
I've been [running some experiments](https://github.com/comfyanonymous/ComfyUI/issues/1992#issuecomment-2306398572) and I reckon this thread is describing the same phenomenon?  Any thoughts on where to go from here?
